File: src/dataset/go/linked_list_Exercism_go_thinkverse/openAI/ChatGPT4_linked_list_v3.go
package linkedlist

type Node struct {
	Value interface{}
	prev  *Node
	next  *Node
}

type List struct {
	head *Node
	tail *Node
	size int
}

func NewList(elements ...interface{}) *List {
	list := &List{}
	for _, el := range elements {
		list.Push(el)
	}
	return list
}

func (n *Node) Next() *Node {
	return n.next
}

func (n *Node) Prev() *Node {
	return n.prev
}

func (l *List) Unshift(v interface{}) {
	newNode := &Node{Value: v}
	if l.head == nil {
		l.head, l.tail = newNode, newNode
	} else {
		newNode.next = l.head
		l.head.prev = newNode
		l.head = newNode
	}
	l.size++
}

func (l *List) Push(v interface{}) {
	newNode := &Node{Value: v}
	if l.tail == nil {
		l.head, l.tail = newNode, newNode
	} else {
		newNode.prev = l.tail
		l.tail.next = newNode
		l.tail = newNode
	}
	l.size++
}

func (l *List) Shift() (interface{}, error) {
	if l.head == nil {
		return nil, nil
	}
	value := l.head.Value
	l.head = l.head.next
	if l.head != nil {
		l.head.prev = nil
	} else {
		l.tail = nil
	}
	l.size--
	return value, nil
}

func (l *List) Pop() (interface{}, error) {
	if l.tail == nil {
		return nil, nil
	}
	value := l.tail.Value
	l.tail = l.tail.prev
	if l.tail != nil {
		l.tail.next = nil
	} else {
		l.head = nil
	}
	l.size--
	return value, nil
}

func (l *List) Reverse() {
	if l.size < 2 {
		return
	}
	current := l.head
	var temp *Node
	l.head, l.tail = l.tail, l.head
	for current != nil {
		temp = current.next
		current.next, current.prev = current.prev, temp
		current = temp
	}
}

func (l *List) First() *Node {
	return l.head
}

func (l *List) Last() *Node {
	return l.tail
}

File: src/dataset/javascript/linked-list_exercism-javascript-shyvum/openAI/ChatGPT4_linked-list_v3.js
export class LinkedList {
  constructor() {
    this.head = { next: null };
    this.tail = { prev: this.head };
    this.head.next = this.tail;
    this.size = 0;
  }
  push(data) {
    this.insert(data, this.tail.prev);
  }
  unshift(data) {
    this.insert(data, this.head);
  }
  pop() {
    return this.size > 0 ? this.del(this.tail.prev) : undefined;
  }
  shift() {
    return this.size > 0 ? this.del(this.head.next) : undefined;
  }
  delete(data) {
    const node = this.find(data);
    if (node) this.del(node);
  }
  count() {
    return this.size;
  }
  find(data) {
    let node = this.head.next;
    while (node !== this.tail) {
      if (node.data === data) return node;
      node = node.next;
    }
    return null;
  }
  insert(data, prev) {
    const next = prev.next;
    const newNode = { data, prev, next };
    prev.next = next.prev = newNode;
    this.size++;
  }
  del(node) {
    node.prev.next = node.next;
    node.next.prev = node.prev;
    this.size--;
    return node.data;
  }
  *nodes() {
    let node = this.head.next;
    while (node !== this.tail) {
      yield node;
      node = node.next;
    }
  }
  firstNode() {
    return this.size > 0 ? this.head.next : null;
  }
  lastNode() {
    return this.size > 0 ? this.tail.prev : null;
  }
}